api/apiclient.go

```diff
@@ -605,13 +601,16 @@ func dialAPI(info *Info, opts0 DialOpts) (*dialResult, error) {
 
 // gorillaDialWebsocket makes a websocket connection using the
 // gorilla websocket package.
-func gorillaDialWebsocket(ctx context.Context, urlStr string, tlsConfig *tls.Config) (jsoncodec.JSONConn, error) {
+func gorillaDialWebsocket(ctx context.Context, urlStr string, tlsConfig *tls.Config, ipAddr string) (jsoncodec.JSONConn, error) {
 	// TODO(rogpeppe) We'd like to set Deadline here
 	// but that would break lots of tests that rely on
 	// setting a zero timeout.
 	netDialer := net.Dialer{}
 	dialer := &websocket.Dialer{
 		NetDial: func(netw, addr string) (net.Conn, error) {
+			if ipAddr != "" {
```

Comment:
We're not replacing DNS - we're just doing the DNS lookup separately so that we can do the dials concurrently. The logic inside net.Dial just tries the addresses sequentially, so if the first IP address refers to a controller that's down or unavailable, the dial might take a very long time to succeed even though another of the IP addresses is immediately available.

We need the TLS logic to understand that it's connecting to the pre-DNS-resolution address, but we want the net dial to go directly to the IP address that we've chosen.